src/user-rank/user-ranks.service.ts

```diff
@@ -0,0 +1,199 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+import { Injectable, NotFoundException } from '@nestjs/common';
+import { User } from '@prisma/client';
+import is from '@sindresorhus/is';
+import { PrismaService } from '../prisma/prisma.service';
+import { ListUsersWithRankOptions } from '../users/interfaces/list-by-rank-options';
+import { SerializedUserWithRank } from '../users/interfaces/serialized-user-with-rank';
+import { EventType } from '.prisma/client';
+
+@Injectable()
+export class UserRankService {
+  constructor(private readonly prisma: PrismaService) {}
+
+  async updateRanks(): Promise<void> {
+    for (const eventType of [...Object.keys(EventType), 'total_points']) {
+      await this.prisma.$executeRawUnsafe(
+        `REFRESH MATERIALIZED VIEW ${eventType}_user_ranks;`,
+      );
+    }
+  }
+
+  async findOrThrow(id: number): Promise<User> {
```

Comment:
We should be able to fix the circular dependency after removing the injection in the users service